src/main/java/com/shapesecurity/bandolier/es2017/transformations/VariableReferenceExtractor.java

```diff
@@ -0,0 +1,90 @@
+package com.shapesecurity.bandolier.es2017.transformations;
+
+import com.shapesecurity.functional.data.Maybe;
+import com.shapesecurity.functional.data.MultiHashTable;
+import com.shapesecurity.shift.es2017.ast.AssignmentTargetIdentifier;
+import com.shapesecurity.shift.es2017.ast.BindingIdentifier;
+import com.shapesecurity.shift.es2017.ast.IdentifierExpression;
+import com.shapesecurity.shift.es2017.ast.Module;
+import com.shapesecurity.shift.es2017.ast.Program;
+import com.shapesecurity.shift.es2017.ast.Script;
+import com.shapesecurity.shift.es2017.reducer.Director;
+import com.shapesecurity.shift.es2017.reducer.MonoidalReducer;
+import com.shapesecurity.shift.es2017.scope.GlobalScope;
+import com.shapesecurity.shift.es2017.scope.ScopeAnalyzer;
+import com.shapesecurity.shift.es2017.scope.ScopeLookup;
+import com.shapesecurity.shift.es2017.scope.Variable;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.stream.Collectors;
+
+// determines all declared global variables for a module, linked to a scope analysis variable
+public class VariableReferenceExtractor {
+
+	private VariableReferenceExtractor() {
+
+	}
+
+	public static MultiHashTable<String, Variable> extractAllReferencedVariables(@Nonnull Program module, @Nonnull GlobalScope globalScope, @Nonnull ScopeLookup lookup) {
+		DeclaredVariableReducer reducer = new DeclaredVariableReducer(globalScope, lookup);
+		return Director.reduceProgram(reducer, module);
+	}
+
+	public static MultiHashTable<String, Variable> extractAllReferencedVariables(@Nonnull Program module) {
+		GlobalScope globalScope = module instanceof Module ? ScopeAnalyzer.analyze((Module) module) : ScopeAnalyzer.analyze((Script) module);
+		return extractAllReferencedVariables(module, globalScope, new ScopeLookup(globalScope));
+	}
+
+	private static final class DeclaredVariableReducer extends MonoidalReducer<MultiHashTable<String, Variable>>  {
```

Comment:
Is there a reason to do this with a reducer, instead of just extracting the relevant information from the Scope object directly?